logstash-core/src/main/java/org/logstash/ackedqueue/QueueUpgrade.java

```diff
@@ -0,0 +1,91 @@
+package org.logstash.ackedqueue;
+
+import com.google.common.primitives.Ints;
+import java.io.File;
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.nio.file.DirectoryStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.logstash.Event;
+import org.logstash.ackedqueue.io.CheckpointIO;
+import org.logstash.ackedqueue.io.FileCheckpointIO;
+import org.logstash.ackedqueue.io.MmapPageIOV1;
+import org.logstash.ackedqueue.io.MmapPageIOV2;
+
+public final class QueueUpgrade {
+
+    private static final Logger LOGGER = LogManager.getLogger(QueueUpgrade.class);
+
+    private static final Pattern PAGE_NAME_PATTERN = Pattern.compile("page\\.\\d+");
+
+    private QueueUpgrade() {
+        // Utility Class.
+    }
+
+    public static void upgradeQueueDirectoryToV2(final Path path) throws IOException {
+        final File upgradeFile = path.resolve(".queue-version").toFile();
+        if (!upgradeFile.exists()) {
+            LOGGER.info("No PQ version file found, upgrading to PQ v2.");
+            try (final DirectoryStream<Path> files = Files.newDirectoryStream(path)) {
+                final Collection<File> oldFiles = new ArrayList<>();
+                files.forEach(file -> {
+                    final Matcher matcher = PAGE_NAME_PATTERN.matcher(file.getFileName().toString());
+                    if (matcher.matches()) {
+                        oldFiles.add(file.toFile());
+                    }
+                });
+                final CheckpointIO cpIo = new FileCheckpointIO(path);
+                for (final File v1PageFile : oldFiles) {
+                    final int num =
+                        Integer.parseInt(v1PageFile.getName().substring("page.".length()));
+                    try (final MmapPageIOV1 iov1 = new MmapPageIOV1(
+                        num, Ints.checkedCast(v1PageFile.length()), path
+                    )) {
+                        final Checkpoint cp = cpIo.read(cpIo.tailFileName(num));
+                        final int count = cp.getElementCount();
+                        final long minSeqNum = cp.getMinSeqNum();
+                        iov1.open(minSeqNum, count);
+                        for (int i = 0; i < count; ++i) {
+                            try {
+                                Event.deserialize(
+                                    iov1.read(minSeqNum + 1L, 1).getElements().get(0)
+                                );
+                            } catch (final IOException ex) {
+                                throw new IllegalStateException(
+                                    "Logstash was unable to upgrade your persistent queue." +
```

Comment:
@tsg @andrewvc almost done cleaning this up, will ping you guys when this is ready for full review (1-2h probably, writing a test is pretty tricky but incoming :)).
Just thought I'd ping you here for the error message. Put this as a placeholder now, but you probably want something nicer (more verbose) here right? :)

Reply:
Looks pretty good to me already. If the command to drain the queue is a one liner, maybe we can add it in brackets after "fully drain your persistent queue". Or is it too long?

Reply:
@tsg it's a little specific, you'd need to run LS with an adjusted config that doesn't have an input section and has the `logstash.yml` configured for `queue.drain=true`.
Maybe we should write a block post and link to that?

Reply:
@original-brownbear why do we need to shutdown the inputs? Shouldn't `queue.drain` continue draining the queue on the shutdown signal? Wouldn't the correct process be to enable the queue drain, then send a shutdown signal?

Reply:
@andrewvc phew in theory yes, but I thought this would be cleaner, compared with asking people to only interrupt once. I'm fine with either approach though, that's why I pinged you guys :)

Reply:
Reading through https://github.com/elastic/logstash/blob/6.2/logstash-core/lib/logstash/pipeline.rb that seems to be the intent of the code, unless there's a bug in it.